File: src/nu/_config/telemetry.py
"""Fire-and-forget PostHog capture, no runtime dependency.

POSTs one event to the PostHog capture endpoint on a daemon thread and
returns immediately. Silent on every failure (offline, host down,
malformed response) so a Nu program never blocks or crashes because of
telemetry.

Skipped when: telemetry is off in ``~/.nu/config.json``, we detect an
editable install (dev mode), or no PostHog token is configured. See
:mod:`nu._config.config`.
"""

from __future__ import annotations

import json
import os
import platform
import sys
import threading
import urllib.request
from typing import TYPE_CHECKING

from . import config as _config


if TYPE_CHECKING:
    from typing import Any


# PostHog Project API key -- public write-only token, safe to embed
# (same one baked into the docs site's client bundle). Env var overrides
# for local testing against a different PostHog project.
_TOKEN = os.environ.get("NU_POSTHOG_TOKEN", "phc_yRpWfpoMeeRM8sKMPGPEsGy8ygNHJTpsGwrsikfpRfFz")
_HOST = os.environ.get("NU_POSTHOG_HOST", "https://t.nustack.dev").rstrip("/")


def _base_properties() -> dict[str, Any]:
    from nu import __version__

    return {
        "nu_version": __version__,
        "python_version": f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
        "platform": platform.system().lower(),
        "arch": platform.machine().lower(),
    }


def _post(payload: dict[str, Any]) -> None:
    req = urllib.request.Request(  # noqa: S310
        f"{_HOST}/i/v0/e/",
        data=json.dumps(payload).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=3) as resp:  # noqa: S310
            resp.read()
    except Exception:  # noqa: S110
        pass


def _fire(event: str, distinct_id: str, properties: dict[str, Any] | None = None) -> None:
    payload = {
        "api_key": _TOKEN,
        "event": event,
        "distinct_id": distinct_id,
        "properties": {**_base_properties(), **(properties or {})},
    }
    threading.Thread(target=_post, args=(payload,), daemon=True).start()


def capture(event: str, properties: dict[str, Any] | None = None) -> None:
    """Queue a PostHog event. Loads config to check opt-out. Never raises."""
    try:
        if not _TOKEN or _config.is_dev_install() or not _config.telemetry_enabled():
            return
        _fire(event, _config.distinct_id(), properties)
    except Exception:  # noqa: S110
        pass


def capture_first_run(distinct_id: str) -> None:
    """Fire ``nu_first_run`` without re-reading config -- caller just wrote it."""
    try:
        if not _TOKEN or _config.is_dev_install():
            return
        _fire("nu_first_run", distinct_id)
    except Exception:  # noqa: S110
        pass


def config_for_browser() -> dict[str, Any]:
    """Payload for the nudle ``/api/telemetry-config`` endpoint."""
    enabled = bool(_TOKEN) and _config.telemetry_enabled() and not _config.is_dev_install()
    return {
        "enabled": enabled,
        "distinct_id": _config.distinct_id(),
        "posthog_token": _TOKEN if enabled else "",
        "posthog_host": _HOST,
        **_base_properties(),
    }

File: pkgs/ts/nudle/src/telemetry.ts
// Anonymous nudle launch ping. Fetches `/api/telemetry-config` from the
// Python backend (source of truth for the ~/.nu/config.json opt-out flag
// and the PostHog token/host), then POSTs one `nudle_opened` event to
// PostHog's capture endpoint directly -- no SDK, no autocapture, no
// background machinery. Skipped in vite dev mode. Never throws.

type TelemetryConfig = {
	enabled: boolean;
	distinct_id: string;
	posthog_token: string;
	posthog_host: string;
	nu_version?: string;
	python_version?: string;
	platform?: string;
	arch?: string;
};

export async function initTelemetry(): Promise<void> {
	if (import.meta.env.DEV) return;
	try {
		const res = await fetch("/api/telemetry-config");
		if (!res.ok) return;
		const cfg = (await res.json()) as TelemetryConfig;
		if (!cfg.enabled || !cfg.posthog_token) return;
		const payload = {
			api_key: cfg.posthog_token,
			event: "nudle_opened",
			distinct_id: cfg.distinct_id,
			properties: {
				nu_version: cfg.nu_version,
				python_version: cfg.python_version,
				platform: cfg.platform,
				arch: cfg.arch,
			},
		};
		await fetch(`${cfg.posthog_host.replace(/\/$/, "")}/i/v0/e/`, {
			method: "POST",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify(payload),
			keepalive: true,
		});
	} catch {
		// swallow: telemetry never breaks the app
	}
}
